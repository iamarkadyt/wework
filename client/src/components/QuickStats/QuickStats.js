import React, { Fragment } from 'react'
import { connect } from 'react-redux'
import {
    FaThumbsUp as IcoLike,
    FaComments as IcoComments
} from 'react-icons/fa'

import { withEither } from '../../hocs/withEither'
import FBSpinner from '../FBSpinner/FBSpinner'
import './QuickStats.css'


/**
 * Layout:
 * Image
 * Name
 * Title at Company
 * Followers
 * Following
 * Post count
 * (?) Likes count
 * Comments count
 */

const Stats = ({
    authedUser: {
        name, avatar, followers, following
    },
    usersProfile: {
        title, company, status
    }
}) => (
        <Fragment>
            <section className="QuickStats-name">
                <img src={avatar} alt=""></img>
                <h2>{name}</h2>
                <p>{title} at {company}</p>
                <p>{status}</p>
            </section>
            <section className="QuickStats-rows">
                <h3>Your stats:</h3>
                <p>
                    <span>Following:</span>
                    <span>{following.length}</span>
                </p>
                <p>
                    <span>Your followers:</span>
                    <span>{followers.length}</span>
                </p>
                <p>
                    <span>Total posts:</span>
                    <span>{10}</span>
                </p>
            </section>
            <section className="QuickStats-likes-comments">
                <h3>Total likes and comments on your posts:</h3>
                <div>
                    <p><IcoLike /> 178</p>
                    <p><IcoComments /> 45</p>
                </div>
            </section>
        </Fragment>
    )

const isProfileUnavailableFn = ({ usersProfile }) => !usersProfile
const StatsWithCondRenderings = withEither(isProfileUnavailableFn, FBSpinner)(Stats)

export default connect(state => ({
    authedUser: state.user,
    usersProfile: state.profile
}))(({ authedUser, usersProfile }) => (
    <div className="QuickStats-container">
        <StatsWithCondRenderings
            authedUser={authedUser}
            usersProfile={usersProfile} />
    </div>
))