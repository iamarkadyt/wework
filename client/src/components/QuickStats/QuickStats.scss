@import 'variables.scss';
.QuickStats-container {
    position: sticky;
    box-shadow: $box_shadow_main;
    background: $color-bg-secondary;
    color: $color-font-main;
    border-radius: .1rem;
    padding: 1rem;
    width: 100%;
    top: 4rem;
    .lds-facebook {
        position: relative;
        left: 50%;
        transform: translateX(-50%);
    }
    section:not(:first-child) {
        padding-top: 1rem;
    }
    .QuickStats-name {
        text-align: center;
        img {
            width: 5rem;
            border-radius: 5rem;
        }
    }
    .QuickStats-rows {
        h3 {
            margin-bottom: .5rem;
        }
        div {
            display: flex;
            justify-content: space-between;
            align-items: center;
            font-size: 1.1rem;
            width: 100%;
            div {
                border-bottom: dotted $color-font-secondary .1rem;
                height: .75rem;
                margin: 0 .5rem;
                flex-shrink: 2000;
            }
        }
    }
    .QuickStats-likes-comments {
        div {
            display: grid;
            grid-template-columns: 1fr 1fr;
            justify-items: center;
            p {
                padding: .5rem .5rem 0 .5rem;
                font-size: 2rem;
            }
        }
    }
}