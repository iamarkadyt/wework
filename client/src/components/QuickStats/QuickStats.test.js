import React from 'react'
import { shallow } from 'enzyme'
import { QuickStats, mapStateToProps } from './QuickStats'
import cloneDeep from 'lodash.clonedeep'

jest.mock('../../hocs/conditionalRendering')
jest.mock('recompose')

const { withEither } = require('../../hocs/conditionalRendering')
const { compose } = require('recompose')

// mocks become fully functional
// this removes need for manual unmocking per every test
// UPDATE: actually I think I could have just used jest.spyOn?
withEither.mockImplementation(
  require.requireActual('../../hocs/conditionalRendering').withEither
)
compose.mockImplementation(
  require.requireActual('recompose').compose
)

let mountedComponent, props

const getMockProps = () => {
  return cloneDeep({
    authedUser: {
      name: 'Mike',
      avatar: undefined
    },
    usersProfile: {
      title: 'Construction Worker',
      company: 'Construction Workers United',
      status: 'Not Open For Employment'
    },
    stats: {
      followers: 1038,
      following: 1270,
      postCount: 20,
      totalLikes: 503,
      totalComments: 121
    },
    history: {
      push: jest.fn()
    },
    fetchUsersStats: jest.fn()
  })
}

const comp = () => {
  if (!mountedComponent) {
    mountedComponent = shallow(<QuickStats {...props} />)
  }
  return mountedComponent
}

describe('QuickStats', () => {
  beforeEach(() => {
    mountedComponent = undefined
    props = getMockProps()
  })

  describe('functions', () => {
    describe('componentDidMount', () => {
      it('calls fetchUsersStats', () => {
        comp()
        expect(props.fetchUsersStats).toHaveBeenCalled()
      })
    })

    describe('render', () => {
      it('calls withCondRendering', () => {
        const mock = jest.spyOn(QuickStats.prototype, 'withCondRendering')
        comp()
        expect(mock).toHaveBeenCalled()
      })
    })

    describe('withCondRendering', () => {
      it('calls withEither', () => {
        withEither.mockClear()
        comp()
        expect(withEither).toHaveBeenCalledTimes(2)
      })

      it('calls compose', () => {
        compose.mockClear()
        comp()
        expect(compose).toHaveBeenCalledTimes(1)
      })
    })

    describe('isLoadingFn', () => {
      it('returns true if stats isn\'t defined', () => {
        delete props.stats
        expect(comp().instance().isLoadingFn(props)).toBe(true)
      })

      it('returns true if profile isn\'t defined', () => {
        delete props.usersProfile
        expect(comp().instance().isLoadingFn(props)).toBe(true)
      })

      it('returns true if authedUser isn\'t defined', () => {
        delete props.authedUser
        expect(comp().instance().isLoadingFn(props)).toBe(true)
      })

      it('returns false if all three (profile, stats, authedUser) are defined', () => {
        props.authedUser = {}
        props.usersProfile = {}
        props.stats = {}
        expect(comp().instance().isLoadingFn(props)).toBe(false)
      })
    })

    describe('hasNoProfile', () => {
      it('returns true if errors.noProfile is defined', () => {
        expect(comp().instance().hasNoProfile({ errors: { noProfile: 'msg' } })).toBe(true)
      })

      it('returns false if errors.noProfile is undefined', () => {
        expect(comp().instance().hasNoProfile({ errors: {} })).toBe(false)
      })
    })
  })

  describe('mapStateToProps', () => {
    it('returns expected object', () => {
      const state = {
        user: { 
          name: 'John Doe', 
          stats: { 
            height: 200 
          } 
        },
        profile: { 
          age: 30 
        },
        err: { 
          email: 'Email is wrong' 
        }
      }

      const expectedObject = {
        authedUser: state.user,
        usersProfile: state.profile,
        stats: state.user.stats,
        errors: state.err
      }

      expect(mapStateToProps(state)).toEqual(expectedObject)
    })
  })

  it('matches snapshot', () => {
    expect(comp()).toMatchSnapshot()
  })
})

