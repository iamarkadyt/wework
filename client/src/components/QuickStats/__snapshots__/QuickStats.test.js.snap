// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuickStats matches snapshot 1`] = `
<div
  className="QuickStats-container"
>
  <Component
    authedUser={
      Object {
        "avatar": undefined,
        "name": "Mike",
      }
    }
    history={
      Object {
        "push": [MockFunction],
      }
    }
    stats={
      Object {
        "followers": 1038,
        "following": 1270,
        "postCount": 20,
        "totalComments": 121,
        "totalLikes": 503,
      }
    }
    usersProfile={
      Object {
        "company": "Construction Workers United",
        "status": "Not Open For Employment",
        "title": "Construction Worker",
      }
    }
  />
</div>
`;
